keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+  vendor-specific claim parameters into a canonical form (i.e. a generic
+  `ResourceClaimParameters` object in the `resource.k8s.io` API group) which
+  the scheduler (or Cluster Autoscaler) can evaluate against the
+  `NodeResourceSlice` of any candidate nodes without knowing exactly what is
+  being requested. They then use this information to help decide which node to
+  schedule a pod on (as well as allocate resources from its `NodeResourceSlice`
+  in the process).
+
+* Once a node is chosen and a workload is sent to it, DRA drivers are
+  responsible for injecting any allocated resources into the Pod, according to
+  the resource choices made by the scheduler. This includes applying any
+  configuration information attached to the vendor-specific claim parameters
+  object used in the request.
+
+In this KEP, we define a single "semantic model" called
+`PartitionableResources`, and use it throughout the discussion to help describe
+the framework as a whole. This model is a natural extension to the API provided
+by the traditional device-plugin, which supports advertising resources as a
+finite set of opaque strings. The major difference being that under the new
+model, one can attach a list of attributes to those opaque strings which the
+scheduler can use to select a particular resource for allocation. Additionally,
+each resource can be divided into a set of pre-defined partitions, each with
+their own set of unique attributes attached to them. In this way, requests for
+resources can be satisfied either by a complete resource instance or by one of
+its partitions.  This model may get enhanced in the future and/or other models
+might get added as needed.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Enable cluster autoscaling when pods use resource claims, with correct
+  decisions and changing the cluster size by more than one node at a time.
+
+- Support node-local resources. Adding or removing nodes has no effect
+  on network-attached resources and therefore CA does not need to (and cannot)
+  simulate them.
+
+- Allow DRA driver developers to provide a user experience that is similar to
+  the one possible without semantic parameters. Ideally, users should not notice
+  at all that a driver is using semantic parameters under the hood.
+
+### Non-Goals
+
+- Scheduling performance is expected to become better compared to using the
+  PodSchedulingContext. However, this is not the reason for this KEP.
+
+
+## Proposal
+
+### Publishing node resources
+
+The kubelet publishes NodeResourceSlices to the API server with content
```

Comment:
My intent at some point was to allow admins to create NodeResourceSlices for pod groups, as information for the autoscaler. @MaciekPytel preferred to have that information in the autoscaler configuration or to get it from cloud providers, so at this point only kubelet is expected to publish these objects.

Either way, kubelet would have been responsible for all objects with the same nodeName field as the node on which kubelet runs. I'll make that clearer:

> The kubelet publishes NodeResourceSlices to the API server with content provided by DRA drivers running on its own node. It's also responsible for deleting or updating stale objects. The `nodeName` field determines which objects each kubelet instance is responsible for.

Reply:
Is the following valid or invalid:

The cloud-provider knows all about the machines and hardware.  They run a controller which creates NodeResourceSlices for each machine instead of kubelet.

I guess, at the end of it all, kubelet has to feed device allocation decisions back to the runtime, so it's sort of moot, but I just hate being over-prescriptive.

Reply:
I think it's valid. Especially considering that these are *per-driver*. So, kubelet and the on-node agents may know about some things, but for some drivers (e.g., vlan / vrf / routing policy configs on the ToR or distribution routers) they may actually not know, and a different agent populates this.